event/src/traits.rs

```diff
@@ -3,82 +3,75 @@ mod private_ {
 
     #[cfg(feature = "crossbeam-channel")]
     impl<T> SealedListen for crate::chans::Listener<T> {}
-    #[cfg(feature = "crossbeam-channel")]
-    impl<T> SealedListen for crate::dchans::Listener<T> {}
-    impl<T> SealedListen for crate::schans::Listener<T> {}
 
     impl<T> SealedListen for crate::nonrc::Listener<'_, T> {}
     impl<T> SealedListen for crate::nonts::Listener<T> {}
+    impl<T> SealedListen for crate::ts::Listener<T> {}
 }
 
 /// Internal glue code to avoid boilerplate and repetition
 pub mod private {
-    use crate::RawEventQueue;
-
-    pub(crate) trait QueueInterface<T>: Default {
-        fn with_inner<F, R>(&self, f: F) -> R
-        where
-            F: FnOnce(&RawEventQueue<T>) -> R;
+    pub trait Listen: super::private_::SealedListen {
+        type Item;
 
-        fn with_inner_mut<F, R>(&self, f: F) -> R
-        where
-            F: FnOnce(&mut RawEventQueue<T>) -> R;
-    }
-
-    pub trait Listen<T> {
         fn with_inner_mut<F, R>(&self, f: F) -> Option<R>
         where
-            F: FnOnce(crate::intern::ListenerKey, &mut RawEventQueue<T>) -> R;
-    }
-
-    #[inline]
-    pub(crate) fn extend<T, EQ, I>(equeue: &EQ, events: I) -> bool
-    where
-        EQ: QueueInterface<T>,
-        I: IntoIterator<Item = T>,
-    {
-        equeue.with_inner_mut(|inner| {
-            if inner.listeners.is_empty() {
-                return false;
-            }
-            inner.events.extend(events);
-            true
-        })
+            F: FnOnce(crate::intern::ListenerKey, &mut crate::RawEventQueue<Self::Item>) -> R;
     }
 }
 
-/// Every supported event queue implements this trait
-pub trait GenericQueueInterface<T> {
-    /// Pushes/emits an event
-    fn push(&self, event: T) -> bool;
+use std::borrow::Cow;
 
-    /// Emits multiple events in one go
-    fn extend<I>(&self, events: I) -> bool
-    where
-        I: IntoIterator<Item = T>,
-    {
-        for i in events.into_iter() {
-            if !self.push(i) {
-                return false;
-            }
-        }
-        true
-    }
+pub trait QueueInterfaceCommon {
+    type Item;
 
     /// Checks if any events are currently buffered
     ///
     /// # Return value
     /// * `false` if events are currently buffered
     /// * `true` if no events are currently buffered or the event queue doesn't
     ///   support querying this information
-    fn is_empty(&self) -> bool {
+    fn buffer_is_empty(&self) -> bool {
         true
     }
 }
 
+/// Every supported event queue implements this trait, mutable variant
+///
+/// More types implement this trait (including all types which implement
+/// [`Emitter`]), but can't be fully accessed via `Rc`/`Arc`.
+///
+/// This trait should be used as a trait bound if possible
+/// (instead of the non-mut variant).
+pub trait EmitterMut: QueueInterfaceCommon
+where
+    Self::Item: Clone,
+{
+    /// Pushes/emits an event
+    fn emit(&mut self, event: Cow<'_, Self::Item>) -> bool;
```

Comment:
Is the `bool` an indication of error? If so then you should probably use `Result`, maybe with the error type as an associated.

Reply:
The `bool` is not necessaryly an error, but indicates that there are no possible receivers for the event (e.g. real delivery, not simple `notification`, e.g. further forwarding of a `unit` `()` or a simle wake up). You're right, perhaps I should replace it with a custom enum (as Result doesn't really fit here).

Reply:
related issue: https://github.com/rust-windowing/winit/issues/1292